Add supported Directives list (#1168)
rust-lang/reference: src/inline-assembly.md

On x86, the `.intel_syntax noprefix` mode of GAS is used by default.
On ARM, the `.syntax unified` mode is used.
These targets impose an additional restriction on the assembly code: any assembler state (e.g. the current section which can be changed with `.section`) must be restored to its original value at the end of the asm string.
Assembly code that does not conform to the GAS syntax will result in assembler-specific behavior. 

> Further constraints on the directives used by the assembly are indicated by [Directives Support](#directives-support). 

Reviewer: ```suggestion
Further constraints on the directives used by inline assembly are indicated by [Directives Support](#directives-support). 
```

  - The compiler is currently unable to detect this due to the way inline assembly is compiled, but may catch and reject this in the future.

> **Note**: As a general rule, the flags covered by `preserves_flags` are those which are *not* preserved when performing a function call.

### Directives Support

> Inline ASM supports a subset of the directives supported by both GNU AS and LLVM's internal assembler, given as follows.

Reviewer: ```suggestion
Inline assembly supports a subset of the directives supported by both GNU AS and LLVM's internal assembler, given as follows.
```

The result of using other directives is assembler-specific (and may cause an error, or may be accepted as-is).

> 

Reviewer: ```suggestion

If inline assembly includes any "stateful" directive that modifies how subsequent assembly is processed, the block must undo the effects of any such directives before the inline assembly ends.

```

The following directives are guaranteed to be supported by the assembler:

> - `.2byte`

Reviewer: Could you please either sort the list, or group the list into logical groups?

It may make some sense to group things like `byte`, `short`, `word`, `quad`, and similar together, but that seems like substantially more organizational work, and in practice I would expect that a fully sorted list may be easier to use as a reference.

- `.4byte`
- `.8byte`
- `.byte`
- `.short`
- `.word`
- `.long`
- `.quad`
- `.float`
- `.double`
- `.octa`
- `.sleb128`
- `.uleb128`
- `.ascii`
- `.asciz`
- `.string`
- `.skip`
- `.space`
- `.balign`
- `.balignl`
- `.balignw`
- `.balign`
- `.balignl`
- `.balignw`
- `.fill`
- `.section`
- `.pushsection`

> - `.popsection`

Reviewer: ```suggestion
- `.popsection`
- `.subsection`
```

- `.text`
- `.bss`
- `.data`
- `.def`
- `.endef`
- `.scl`
- `.comm`
- `.lcomm`
- `.option`

> - `.equ`

Reviewer: ```suggestion
- `.equ`
- `.equiv`
- `.eqv`
```

- `.set`
- `.align`
- `.inst`
- `.p2align`

> 

Reviewer: ```suggestion
- `.bundle_align_mode`
- `.bundle_lock`
- `.bundle_unlock`
```

Reviewer: Also, directives like `.dc`, `.dcb`, and `.ds`. Those simply emit data; if they're supported by LLVM we should include them.

> The following directives are guaranteed to be supported for `global_asm` only:

Reviewer: I would prefer to unify these lists, and not suggest that some directives will work within `global_asm` but not within `asm`.

- `.alt_entry`
- `.private_extern`
- `.globl`
- `.global`
- `.size`
- `.type`

#### Target Specific Directive Support

##### Dwarf Unwinding

The following directives are supported on ELF targets that support DWARF unwind info:

- `.cfi_adjust_cfa_offset`
- `.cfi_def_cfa`
- `.cfi_def_cfa_offset`
- `.cfi_def_cfa_register`
- `.cfi_endproc`
- `.cfi_escape`
- `.cfi_lsda`
- `.cfi_offset`
- `.cfi_personality`
- `.cfi_register`
- `.cfi_rel_offset`
- `.cfi_remember_state`
- `.cfi_restore`
- `.cfi_restore_state`
- `.cfi_return_column`
- `.cfi_same_value`
- `.cfi_sections`
- `.cfi_signal_frame`
- `.cfi_startproc`
- `.cfi_undefined`
- `.cfi_window_save`

> 

Reviewer: There are several CFI directives not listed here, such as:
- `.cfi_personality_id`
- `.cfi_fde_data`
- `.cfi_inline_ldsa`
(and more, those are just the first three I noticed when reading https://sourceware.org/binutils/docs/as/CFI-directives.html )

Are those verifiably *not* supported by LLVM, or are they undocumented?

Reviewer: Not critical, we can always add these later.

##### Structured Exception Handling

On targets with structured exception Handling, the following additional directives are guaranteed to be supported:

- `.seh_endproc`
- `.seh_endprologue`
- `.seh_proc`
- `.seh_pushreg`
- `.seh_savereg`
- `.seh_setframe`
- `.seh_stackalloc`

> ##### x86

Reviewer: ```suggestion
##### x86 (32-bit and 64-bit)
```

> On x86, the following additional directives are guaranteed to be supported:

Reviewer: ```suggestion
On x86 targets, both 32-bit and 64-bit, the following additional directives are guaranteed to be supported:
```

- `.nops`

> On x86 for `global_asm!` only, the following additional directives are guaranteed to be supported (it is unspecified whether `.code16` or `.code32` are supported for `asm!()`):

Reviewer: This should not be unspecified; they should be valid within `asm` as well.

- `.code16`
- `.code32`

> 

Reviewer: ```suggestion

The directives `.att_syntax` and `.intel_syntax` (including options such as `noprefix`) are not fully supported. Using these directives to affect literal assembly syntax with no operands will work, as long as the state is restored before the end of the assembly block. Using these directives to affect assembly that includes operands is not defined.

```

Reviewer: You already have to support both syntaxes, given the att_syntax option.

I agree that the state restoration is covered above, but I felt that it couldn't hurt to remind people that this in particular is something stateful.

> ##### ARM

Reviewer: This should specify whether it applies to ARM, or to both ARM and ARM64. (I'm guessing some of these directives aren't valid on ARM64.)

Reviewer: cc @Amanieu Any idea here?

Reviewer 2: Yea these are specific to ARM32. Also there are [a lot more directives](https://sourceware.org/binutils/docs/as/ARM-Directives.html), the ones listed here are just ones I happen to be using.

Reviewer: Submitted a suggestion (below) to that effect.

> On ARM for `global_asm!` only, the following additional directives are guaranteed to be supported:

Reviewer: As above, I'd prefer to see these lists unified.

- `.code`
- `.thumb`
- `.thumb_func`

On ARM, the following additional directives are guaranteed to be supported:

- `.fnstart`
- `.fnend`
- `.save`

> - `.movsp`

Reviewer: ```suggestion
- `.movsp`
- `.even`
```

- `.string`
- `.text`
- `.uleb128`
- `.word`

The following directives are guaranteed to be supported for `global_asm` only:

- `.alt_entry`
- `.globl`
- `.global`
- `.private_extern`
- `.size`
- `.type`

#### Target Specific Directive Support

##### Dwarf Unwinding

The following directives are supported on ELF targets that support DWARF unwind info:

- `.cfi_adjust_cfa_offset`
- `.cfi_def_cfa`
- `.cfi_def_cfa_offset`
- `.cfi_def_cfa_register`
- `.cfi_endproc`
- `.cfi_escape`
- `.cfi_lsda`
- `.cfi_offset`
- `.cfi_personality`
- `.cfi_register`
- `.cfi_rel_offset`
- `.cfi_remember_state`
- `.cfi_restore`
- `.cfi_restore_state`
- `.cfi_return_column`
- `.cfi_same_value`
- `.cfi_sections`
- `.cfi_signal_frame`
- `.cfi_startproc`
- `.cfi_undefined`
- `.cfi_window_save`

##### Structured Exception Handling

On targets with structured exception Handling, the following additional directives are guaranteed to be supported:

- `.seh_endproc`
- `.seh_endprologue`
- `.seh_proc`
- `.seh_pushreg`
- `.seh_savereg`
- `.seh_setframe`
- `.seh_stackalloc`

##### x86

On x86, the following additional directives are guaranteed to be supported:
- `.att_syntax`
- `.intel_syntax`
- `.nops`

> Use of the `.att_syntax` and `.intel_syntax` with no parameters is supported, but the syntax must be restored to the option at entry (`.intel_syntax` without the `att_syntax` asm option, or `.att_syntax` with that option) or the behaviour is undefined (the output of the compiler may be corrupted as a result). Use of `.att_syntax` and `.intel_syntax` with an option (such as `.intel_syntax prefix` or `.att_syntax noprefix`) is unsupported and results in assembler-dependant behaviour. If operand interpolations are used between setting the syntax mode with one of these directives, and restoring it to the block's default, the behaviour is undefined.

Author: > If operand interpolations are used between setting the syntax mode with one of these directives, and restoring it to the block's default, the behaviour is undefined.

This could probably easily be unspecified, but depending on how interpolations are implemented, the results could vary wildly.

> Use of the `.att_syntax` and `.intel_syntax` with no parameters is supported, but the syntax must be restored to the option at entry (`.intel_syntax` without the `att_syntax` asm option, or `.att_syntax` with that option) or the behaviour is undefined (the output of the compiler may be corrupted as a result). Use of `.att_syntax` and `.intel_syntax` with an option (such as `.intel_syntax prefix` or `.att_syntax noprefix`) is unsupported and results in assembler-dependant behaviour. If operand interpolations are used between setting the syntax mode with one of these directives, and restoring it to the block's default, the behaviour is undefined.

Reviewer: Minor tweaks, notably allowing parameters equivalent to the default and only disallowing non-default parameters.

```suggestion
Use of the `.att_syntax` and `.intel_syntax` directives with no parameters (or with parameters equivalent to the defaults) is supported, but the syntax must be restored to the option at entry (`.intel_syntax` without the `att_syntax` asm option, or `.att_syntax` with that option) or the behavior is undefined (the compiled output may be corrupted as a result). Use of `.att_syntax` and `.intel_syntax` with a non-default option (such as `.intel_syntax prefix` or `.att_syntax noprefix`) is unsupported. If operand interpolations are used between setting the syntax mode with one of these directives, and restoring it to the block's default, the behaviour is undefined.
```

Author: It may be nice to say an implementation just needs to support `.intel_syntax`/`.att_syntax` and nothing after that. This isn't a huge deal, though, it's just a note.

Reviewer: In practice, handling a single fixed optional parameter that can only take a single value or be omitted entirely, and which has no semantic effect, doesn't seem like it'd be difficult.

I can easily imagine a compiler detecting them and complaining, but if it has the capability to do so, it should detect the whole directive instead and lint towards `options(...)`. I don't think the argument in particular makes that any harder.

On x86 for `global_asm!` only, the following additional directives are guaranteed to be supported (it is unspecified whether `.code16` or `.code32` are supported for `asm!()`):

- `.code16`
- `.code32`

> ##### ARM

Reviewer: ```suggestion
##### ARM (32-bit)
```
